Sources/NIO/Codec.swift

```diff
@@ -38,6 +38,13 @@ public enum DecodingState {
 /// If you move the reader index forward, either manually or by using one of `buffer.read*` methods, you must ensure
 /// that you no longer need to see those bytes again as they will not be returned to you the next time `decode` is called.
 /// If you still need those bytes to come back, consider taking a local copy of buffer inside the function to perform your read operations on.
+///
+/// The `ByteBuffer` passed in as `buffer` is a slice of a larger buffer owned by the `ByteToMessageDecoder` implementation. Some aspects of this buffer are preserved across calls to `decode`, meaning that any changes to those properties you make in your `decode` method will be reflected in the next call to decode. In particular, the following operations are have the described effects:
+
+/// 1. Moving the reader index forward persists across calls. When your method returns, if the reader index has advanced, those bytes are considered "consumed" and will not be available in future calls to `decode`.
+///    Please note, however, that the numerical value of the `readerIndex` itself is not preserved, and may not be the same from one call to the next. Please do not rely on this numerical value: if you need
+///    to recall where a byte is relative to the `readerIndex`, use an offset rather than an absolute value.
+/// 2. Mutating the bytes in the buffer will cause a `fatalError` and so is not allowed. You are only allowed to move the readerIndex forward or consume from the buffer.
```

Comment:
should we say 'mutating the bytes or the readerIndex will cause undefined behaviour and likely crash your program' or something?